lib/products.go

```diff
@@ -0,0 +1,151 @@
+	ArchivePrefix         string
+	PublicKeyId           string
+	PublicKeyUrl          string
+}
+
+type TerraformProduct struct {
+	ProductDetails
+}
+type OpenTofuProduct struct {
+	ProductDetails
+}
+
+type Product interface {
+	GetId() string
+	GetName() string
+	GetDefaultMirrorUrl() string
+	GetVersionPrefix() string
+	GetExecutableName() string
+	GetArchivePrefix() string
+	GetPublicKeyId() string
+	GetPublicKeyUrl() string
+	GetShaSignatureSuffix() string
+	GetArtifactUrl(mirrorURL string, version string) string
+}
+
+// Terraform Product
+func (p TerraformProduct) GetId() string {
+	return p.ID
+}
+func (p TerraformProduct) GetName() string {
+	return p.Name
+}
+func (p TerraformProduct) GetDefaultMirrorUrl() string {
+	return p.DefaultMirror
+}
+func (p TerraformProduct) GetVersionPrefix() string {
+	return p.VersionPrefix
+}
+func (p TerraformProduct) GetExecutableName() string {
+	return p.ExecutableName
+}
+func (p TerraformProduct) GetArchivePrefix() string {
+	return p.ArchivePrefix
+}
+func (p TerraformProduct) GetArtifactUrl(mirrorURL string, version string) string {
+	return fmt.Sprintf("%s%s", mirrorURL, version)
+}
+func (p TerraformProduct) GetPublicKeyId() string {
+	return p.PublicKeyId
+}
+func (p TerraformProduct) GetPublicKeyUrl() string {
+	return p.PublicKeyUrl
+}
+func (p TerraformProduct) GetShaSignatureSuffix() string {
+	return p.GetPublicKeyId() + ".sig"
+}
+
+// OpenTofu methods
+func (p OpenTofuProduct) GetId() string {
+	return p.ID
+}
+func (p OpenTofuProduct) GetName() string {
+	return p.Name
+}
+func (p OpenTofuProduct) GetDefaultMirrorUrl() string {
+	return p.DefaultMirror
+}
+func (p OpenTofuProduct) GetVersionPrefix() string {
+	return p.VersionPrefix
+}
+func (p OpenTofuProduct) GetExecutableName() string {
+	return p.ExecutableName
+}
+func (p OpenTofuProduct) GetArchivePrefix() string {
+	return p.ArchivePrefix
+}
+func (p OpenTofuProduct) GetArtifactUrl(mirrorURL string, version string) string {
+	return fmt.Sprintf("%s/v%s", p.DefaultDownloadMirror, version)
+}
+func (p OpenTofuProduct) GetPublicKeyId() string {
+	return p.PublicKeyId
+}
+func (p OpenTofuProduct) GetPublicKeyUrl() string {
+	return p.PublicKeyUrl
+}
+func (p OpenTofuProduct) GetShaSignatureSuffix() string {
+	return "gpgsig"
+}
+
+// Factory methods
+var products = []Product{
+	TerraformProduct{
+		ProductDetails{
+			ID:             "terraform",
+			Name:           "Terraform",
+			DefaultMirror:  "https://releases.hashicorp.com/terraform",
+			VersionPrefix:  "terraform_",
+			ExecutableName: "terraform",
+			ArchivePrefix:  "terraform_",
+			PublicKeyId:    "72D7468F",
+			PublicKeyUrl:   "https://www.hashicorp.com/.well-known/pgp-key.txt",
+		},
+	},
+	OpenTofuProduct{
+		ProductDetails{
+			ID:                    "opentofu",
+			Name:                  "OpenTofu",
+			DefaultMirror:         "https://get.opentofu.org/tofu",
+			DefaultDownloadMirror: "https://github.com/opentofu/opentofu/releases/download",
```

Comment:
This cannot currently be overriden.
If we do allow it, it's not (currently) used by Terraform.
But, if it is, then the "default download mirror" URL would need to default to releases.hashicorp.com for Terraform and Github got OpenTofu.
However, since they have different layouts:
 * terraform/1.0.0/terraform_1.0x
 * tags/v1.0.0/tofu/
We'd need to make it clear how someone would create their own mirror if they wish to host them!

Reply:
There is a --mirror option with the current version.
Please double check to see if the `--mirror` is still usable after this upgrade

Reply:
@warrensbox It is not currently usable for the actual download source of OpenTofu - it will currently always use Github